NullPointerException in TracingExecutionInterceptor during retries

Hey 👋,

Using `TracingExecutionInterceptor` with `SqsAsyncClient` can throw a NPE during retries when there is a HTTP 500 response.

`TracingExecutionInterceptor` throws NPE in `beforeUnmarshalling()` when `clientSpan` is null. This happens during retries as `modifyHttpRequest()` is only called once initially to setup the `clientSpan`. Further retry calls do not hit `modifyHttpRequest()`, which means once `beforeUnmarshalling()` clears `clientSpan`, the consequent retry responses have a null `clientSpan`.

[Here is a sandbox repo and test to help you reproduce the error.](https://github.com/sguillope/brave-aws-sqs-instrumentation-npe/blob/master/src/test/java/com/example/braveawssqsinstrumentationnpe/BraveAwsSqsInstrumentationNpeApplicationTests.java)

Stack trace:
```
java.util.concurrent.CompletionException: software.amazon.awssdk.core.exception.SdkClientException: Unable to execute HTTP request: null

	at software.amazon.awssdk.utils.CompletableFutureUtils.errorAsCompletionException(CompletableFutureUtils.java:60)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncExecutionFailureExceptionReportingStage.lambda$execute$0(AsyncExecutionFailureExceptionReportingStage.java:51)
	at java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:836)
	at java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:811)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488)
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1990)
	at software.amazon.awssdk.utils.CompletableFutureUtils.lambda$forwardExceptionTo$0(CompletableFutureUtils.java:74)
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:774)
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:750)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488)
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1990)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryingExecutor.maybeAttemptExecute(AsyncRetryableStage.java:85)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryingExecutor.maybeRetryExecute(AsyncRetryableStage.java:144)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryingExecutor.lambda$attemptExecute$1(AsyncRetryableStage.java:125)
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:774)
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:750)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488)
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1990)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeAsyncHttpRequestStage$WrappedErrorForwardingResponseHandler.onError(MakeAsyncHttpRequestStage.java:122)
	at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler.lambda$exceptionCaught$0(ResponseHandler.java:137)
	at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler.runAndLogError(ResponseHandler.java:171)
	at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler.exceptionCaught(ResponseHandler.java:137)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:300)
	at io.netty.channel.AbstractChannelHandlerContext.notifyHandlerException(AbstractChannelHandlerContext.java:825)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:355)
	at software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerPublisher.channelRead(HandlerPublisher.java:396)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:355)
	at software.amazon.awssdk.http.nio.netty.internal.nrs.HttpStreamsHandler.channelRead(HttpStreamsHandler.java:196)
	at software.amazon.awssdk.http.nio.netty.internal.nrs.HttpStreamsClientHandler.channelRead(HttpStreamsClientHandler.java:173)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:355)
	at software.amazon.awssdk.http.nio.netty.internal.LastHttpContentHandler.channelRead(LastHttpContentHandler.java:43)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:355)
	at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:227)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:355)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:355)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:321)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:308)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:422)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276)
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:355)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:355)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.lang.Thread.run(Thread.java:748)
Caused by: software.amazon.awssdk.core.exception.SdkClientException: Unable to execute HTTP request: null
	at software.amazon.awssdk.core.exception.SdkClientException$BuilderImpl.build(SdkClientException.java:98)
	at software.amazon.awssdk.core.exception.SdkClientException.create(SdkClientException.java:43)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.utils.RetryableStageHelper.setLastException(RetryableStageHelper.java:201)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryingExecutor.maybeRetryExecute(AsyncRetryableStage.java:143)
	... 60 more
Caused by: java.lang.NullPointerException
	at brave.instrumentation.awsv2.TracingExecutionInterceptor.beforeUnmarshalling(TracingExecutionInterceptor.java:126)
	at software.amazon.awssdk.core.interceptor.ExecutionInterceptorChain.lambda$beforeUnmarshalling$8(ExecutionInterceptorChain.java:155)
	at software.amazon.awssdk.core.interceptor.ExecutionInterceptorChain.reverseForEach(ExecutionInterceptorChain.java:217)
	at software.amazon.awssdk.core.interceptor.ExecutionInterceptorChain.beforeUnmarshalling(ExecutionInterceptorChain.java:155)
	at software.amazon.awssdk.core.internal.http.async.AsyncAfterTransmissionInterceptorCallingResponseHandler.beforeUnmarshalling(AsyncAfterTransmissionInterceptorCallingResponseHandler.java:57)
	at software.amazon.awssdk.core.internal.http.async.AsyncAfterTransmissionInterceptorCallingResponseHandler.onHeaders(AsyncAfterTransmissionInterceptorCallingResponseHandler.java:67)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeAsyncHttpRequestStage$WrappedErrorForwardingResponseHandler.onHeaders(MakeAsyncHttpRequestStage.java:112)
	at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler.channelRead0(ResponseHandler.java:89)
	at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler.channelRead0(ResponseHandler.java:66)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377)
	... 48 more
```

--- a/brave-instrumentation/aws-java-sdk-v2-core/src/main/java/brave/instrumentation/awsv2/TracingExecutionInterceptor.java
+++ b/brave-instrumentation/aws-java-sdk-v2-core/src/main/java/brave/instrumentation/awsv2/TracingExecutionInterceptor.java
@@ -74,12 +74,18 @@ final class TracingExecutionInterceptor implements ExecutionInterceptor {
     return request.build();
   }
 
+  /**
+   * Before sending an http request. Will be called multiple times in the case of retries.
+   */
   @Override public void beforeTransmission(Context.BeforeTransmission context,
       ExecutionAttributes executionAttributes) {
     Span span = executionAttributes.getAttribute(SPAN);
     span.annotate("ws");
   }
 
+  /**
+   * After sending an http request. Will be called multiple times in the case of retries.
+   */
   @Override public void afterTransmission(Context.AfterTransmission context,
       ExecutionAttributes executionAttributes) {
     Span span = executionAttributes.getAttribute(SPAN);

--- a/brave-instrumentation/aws-java-sdk-v2-core/src/main/java/brave/instrumentation/awsv2/TracingExecutionInterceptor.java
+++ b/brave-instrumentation/aws-java-sdk-v2-core/src/main/java/brave/instrumentation/awsv2/TracingExecutionInterceptor.java
@@ -77,7 +77,8 @@ final class TracingExecutionInterceptor implements ExecutionInterceptor {
   }
 
   /**
-   * Before an individual http request happens, may be run multiple times if retries occur
+   * Before an individual http request is finalized. This is only called once per operation, meaning
+   * we can only have one span per operation.
    */
   @Override public SdkHttpRequest modifyHttpRequest(
       Context.ModifyHttpRequest context,

--- a/brave-instrumentation/aws-java-sdk-v2-core/src/main/java/brave/instrumentation/awsv2/TracingExecutionInterceptor.java
+++ b/brave-instrumentation/aws-java-sdk-v2-core/src/main/java/brave/instrumentation/awsv2/TracingExecutionInterceptor.java
@@ -65,11 +65,10 @@ final class TracingExecutionInterceptor implements ExecutionInterceptor {
 
     String serviceName = executionAttributes.getAttribute(SdkExecutionAttribute.SERVICE_NAME);
     String operation = getAwsOperationNameFromRequestClass(context.request());
-    span.name("aws-sdk")
+    span.name(operation)
+        .remoteServiceName(serviceName)
         .tag("aws.service_name", serviceName)
         .tag("aws.operation", operation);
-
-    span.name(operation).remoteServiceName(serviceName);
 
     return request.build();
   }
@@ -80,6 +79,10 @@ final class TracingExecutionInterceptor implements ExecutionInterceptor {
   @Override public void beforeTransmission(Context.BeforeTransmission context,
       ExecutionAttributes executionAttributes) {
     Span span = executionAttributes.getAttribute(SPAN);
+    if (span == null) {
+      // An evil interceptor deleted our attribute.
+      return;
+    }
     span.annotate("ws");
   }
 
@@ -89,6 +92,10 @@ final class TracingExecutionInterceptor implements ExecutionInterceptor {
   @Override public void afterTransmission(Context.AfterTransmission context,
       ExecutionAttributes executionAttributes) {
     Span span = executionAttributes.getAttribute(SPAN);
+    if (span == null) {
+      // An evil interceptor deleted our attribute.
+      return;
+    }
     span.annotate("wr");
   }
 
@@ -100,6 +107,10 @@ final class TracingExecutionInterceptor implements ExecutionInterceptor {
       ExecutionAttributes executionAttributes
   ) {
     Span span = executionAttributes.getAttribute(SPAN);
+    if (span == null) {
+      // An evil interceptor deleted our attribute.
+      return;
+    }
     handler.handleReceive(new HttpClientResponse(context.httpResponse()), null, span);
   }
 
@@ -111,6 +122,10 @@ final class TracingExecutionInterceptor implements ExecutionInterceptor {
       ExecutionAttributes executionAttributes
   ) {
     Span span = executionAttributes.getAttribute(SPAN);
+    if (span == null) {
+      // An evil interceptor deleted our attribute.
+      return;
+    }
     handler.handleReceive(null, context.exception(), span);
   }
 

--- a/brave-instrumentation/aws-java-sdk-v2-core/src/main/java/brave/instrumentation/awsv2/TracingExecutionInterceptor.java
+++ b/brave-instrumentation/aws-java-sdk-v2-core/src/main/java/brave/instrumentation/awsv2/TracingExecutionInterceptor.java
@@ -48,8 +48,6 @@ final class TracingExecutionInterceptor implements ExecutionInterceptor {
       new ExecutionAttribute<>("DEFERRED_ROOT_CONTEXT");
   static final ExecutionAttribute<Span> APPLICATION_SPAN =
       new ExecutionAttribute<>("APPLICATION_SPAN");
-  static final ExecutionAttribute<Span> CLIENT_SPAN =
-      new ExecutionAttribute<>(Span.class.getCanonicalName());
 
   final Tracer tracer;
   final HttpTracing httpTracing;
@@ -86,48 +84,41 @@ final class TracingExecutionInterceptor implements ExecutionInterceptor {
       ExecutionAttributes executionAttributes
   ) {
     TraceContext maybeDeferredRootSpan = executionAttributes.getAttribute(DEFERRED_ROOT_CONTEXT);
-    Span applicationSpan;
+    Span span;
     HttpClientRequest request = new HttpClientRequest(context.httpRequest());
     if (maybeDeferredRootSpan != null) {
       Boolean sampled = httpTracing.clientRequestSampler().trySample(request);
       if (sampled == null) {
         sampled = httpTracing.tracing().sampler().isSampled(maybeDeferredRootSpan.traceId());
       }
-      applicationSpan = tracer.toSpan(maybeDeferredRootSpan.toBuilder().sampled(sampled).build());
-      executionAttributes.putAttribute(APPLICATION_SPAN, applicationSpan.start());
+      span = tracer.toSpan(maybeDeferredRootSpan.toBuilder().sampled(sampled).build());
+      executionAttributes.putAttribute(APPLICATION_SPAN, span.start());
     } else {
-      applicationSpan = executionAttributes.getAttribute(APPLICATION_SPAN);
+      span = executionAttributes.getAttribute(APPLICATION_SPAN);
     }
 
     String serviceName = executionAttributes.getAttribute(SdkExecutionAttribute.SERVICE_NAME);
     String operation = getAwsOperationNameFromRequestClass(context.request());
-    applicationSpan.name("aws-sdk")
+    span.name("aws-sdk")
         .tag("aws.service_name", serviceName)
         .tag("aws.operation", operation);
 
-    Span span = tracer.newChild(applicationSpan.context());
     handler.handleSend(request, span);
     span.name(operation).remoteServiceName(serviceName);
-    executionAttributes.putAttribute(CLIENT_SPAN, span);
 
     return request.build();
   }
 
-  /**
-   * After individual http response, may run multiple times if retries occur
-   */
-  @Override
-  public void beforeUnmarshalling(
-      Context.BeforeUnmarshalling context,
-      ExecutionAttributes executionAttributes
-  ) {
-    Span clientSpan = executionAttributes.getAttribute(CLIENT_SPAN);
-    if (!context.httpResponse().isSuccessful()) {
-      clientSpan.tag("error", context.httpResponse().statusText()
-          .orElse("Unknown AWS service error"));
-    }
-    handler.handleReceive(new HttpClientResponse(context.httpResponse()), null, clientSpan);
-    executionAttributes.putAttribute(CLIENT_SPAN, null);
+  @Override public void beforeTransmission(Context.BeforeTransmission context,
+      ExecutionAttributes executionAttributes) {
+    Span span = executionAttributes.getAttribute(APPLICATION_SPAN);
+    span.annotate("ws");
+  }
+
+  @Override public void afterTransmission(Context.AfterTransmission context,
+      ExecutionAttributes executionAttributes) {
+    Span span = executionAttributes.getAttribute(APPLICATION_SPAN);
+    span.annotate("wr");
   }
 
   /**
@@ -137,8 +128,9 @@ final class TracingExecutionInterceptor implements ExecutionInterceptor {
       Context.AfterExecution context,
       ExecutionAttributes executionAttributes
   ) {
-    Span applicationSpan = executionAttributes.getAttribute(APPLICATION_SPAN);
-    applicationSpan.finish();
+    Span span = executionAttributes.getAttribute(APPLICATION_SPAN);
+    handler.handleReceive(new HttpClientResponse(context.httpResponse()), null, span);
+    span.finish();
   }
 
   /**
@@ -148,14 +140,10 @@ final class TracingExecutionInterceptor implements ExecutionInterceptor {
       Context.FailedExecution context,
       ExecutionAttributes executionAttributes
   ) {
-    Span clientSpan = executionAttributes.getAttribute(CLIENT_SPAN);
-    if (clientSpan != null) {
-      handler.handleReceive(null, context.exception(), clientSpan);
-      executionAttributes.putAttribute(CLIENT_SPAN, null);
-    }
-    Span applicationSpan = executionAttributes.getAttribute(APPLICATION_SPAN);
-    applicationSpan.error(context.exception());
-    applicationSpan.finish();
+    Span span = executionAttributes.getAttribute(APPLICATION_SPAN);
+    handler.handleReceive(null, context.exception(), span);
+    span.error(context.exception());
+    span.finish();
   }
 
   private String getAwsOperationNameFromRequestClass(SdkRequest request) {

--- a/brave-instrumentation/aws-java-sdk-v2-core/src/main/java/brave/instrumentation/awsv2/TracingExecutionInterceptor.java
+++ b/brave-instrumentation/aws-java-sdk-v2-core/src/main/java/brave/instrumentation/awsv2/TracingExecutionInterceptor.java
@@ -14,12 +14,10 @@
 package brave.instrumentation.awsv2;
 
 import brave.Span;
-import brave.Tracer;
 import brave.http.HttpClientHandler;
 import brave.http.HttpTracing;
 import brave.instrumentation.awsv2.AwsSdkTracing.HttpClientRequest;
 import brave.instrumentation.awsv2.AwsSdkTracing.HttpClientResponse;
-import brave.propagation.TraceContext;
 import software.amazon.awssdk.core.SdkRequest;
 import software.amazon.awssdk.core.interceptor.Context;
 import software.amazon.awssdk.core.interceptor.ExecutionAttribute;
@@ -44,36 +42,12 @@ import software.amazon.awssdk.http.SdkHttpRequest;
  * with the error. The AWS request ID is added when available.
  */
 final class TracingExecutionInterceptor implements ExecutionInterceptor {
-  static final ExecutionAttribute<TraceContext> DEFERRED_ROOT_CONTEXT =
-      new ExecutionAttribute<>("DEFERRED_ROOT_CONTEXT");
-  static final ExecutionAttribute<Span> APPLICATION_SPAN =
-      new ExecutionAttribute<>("APPLICATION_SPAN");
+  static final ExecutionAttribute<Span> SPAN = new ExecutionAttribute<>(Span.class.getName());
 
-  final Tracer tracer;
-  final HttpTracing httpTracing;
   final HttpClientHandler<brave.http.HttpClientRequest, brave.http.HttpClientResponse> handler;
 
   TracingExecutionInterceptor(HttpTracing httpTracing) {
-    this.httpTracing = httpTracing;
-    this.tracer = httpTracing.tracing().tracer();
     this.handler = HttpClientHandler.create(httpTracing);
-  }
-
-  /**
-   * Before the SDK request leaves the calling thread
-   */
-  @Override public void beforeExecution(
-      Context.BeforeExecution context,
-      ExecutionAttributes executionAttributes
-  ) {
-    Span maybeDeferredRootSpan = tracer.nextSpan();
-    if (maybeDeferredRootSpan.context().parentIdAsLong() == 0) {
-      // Deferred to sampling when we have http context
-      // We will build a new span with this context later when we can make the sampling decision
-      executionAttributes.putAttribute(DEFERRED_ROOT_CONTEXT, maybeDeferredRootSpan.context());
-    } else {
-      executionAttributes.putAttribute(APPLICATION_SPAN, maybeDeferredRootSpan.start());
-    }
   }
 
   /**
@@ -84,19 +58,10 @@ final class TracingExecutionInterceptor implements ExecutionInterceptor {
       Context.ModifyHttpRequest context,
       ExecutionAttributes executionAttributes
   ) {
-    TraceContext maybeDeferredRootSpan = executionAttributes.getAttribute(DEFERRED_ROOT_CONTEXT);
-    Span span;
     HttpClientRequest request = new HttpClientRequest(context.httpRequest());
-    if (maybeDeferredRootSpan != null) {
-      Boolean sampled = httpTracing.clientRequestSampler().trySample(request);
-      if (sampled == null) {
-        sampled = httpTracing.tracing().sampler().isSampled(maybeDeferredRootSpan.traceId());
-      }
-      span = tracer.toSpan(maybeDeferredRootSpan.toBuilder().sampled(sampled).build());
-      executionAttributes.putAttribute(APPLICATION_SPAN, span.start());
-    } else {
-      span = executionAttributes.getAttribute(APPLICATION_SPAN);
-    }
+
+    Span span = handler.handleSend(request);
+    executionAttributes.putAttribute(SPAN, span);
 
     String serviceName = executionAttributes.getAttribute(SdkExecutionAttribute.SERVICE_NAME);
     String operation = getAwsOperationNameFromRequestClass(context.request());
@@ -104,7 +69,6 @@ final class TracingExecutionInterceptor implements ExecutionInterceptor {
         .tag("aws.service_name", serviceName)
         .tag("aws.operation", operation);
 
-    handler.handleSend(request, span);
     span.name(operation).remoteServiceName(serviceName);
 
     return request.build();
@@ -112,13 +76,13 @@ final class TracingExecutionInterceptor implements ExecutionInterceptor {
 
   @Override public void beforeTransmission(Context.BeforeTransmission context,
       ExecutionAttributes executionAttributes) {
-    Span span = executionAttributes.getAttribute(APPLICATION_SPAN);
+    Span span = executionAttributes.getAttribute(SPAN);
     span.annotate("ws");
   }
 
   @Override public void afterTransmission(Context.AfterTransmission context,
       ExecutionAttributes executionAttributes) {
-    Span span = executionAttributes.getAttribute(APPLICATION_SPAN);
+    Span span = executionAttributes.getAttribute(SPAN);
     span.annotate("wr");
   }
 
@@ -129,9 +93,8 @@ final class TracingExecutionInterceptor implements ExecutionInterceptor {
       Context.AfterExecution context,
       ExecutionAttributes executionAttributes
   ) {
-    Span span = executionAttributes.getAttribute(APPLICATION_SPAN);
+    Span span = executionAttributes.getAttribute(SPAN);
     handler.handleReceive(new HttpClientResponse(context.httpResponse()), null, span);
-    span.finish();
   }
 
   /**
@@ -141,10 +104,8 @@ final class TracingExecutionInterceptor implements ExecutionInterceptor {
       Context.FailedExecution context,
       ExecutionAttributes executionAttributes
   ) {
-    Span span = executionAttributes.getAttribute(APPLICATION_SPAN);
+    Span span = executionAttributes.getAttribute(SPAN);
     handler.handleReceive(null, context.exception(), span);
-    span.error(context.exception());
-    span.finish();
   }
 
   private String getAwsOperationNameFromRequestClass(SdkRequest request) {

--- a/brave-instrumentation/aws-java-sdk-v2-core/src/test/java/brave/instrumentation/awsv2/ITTracingExecutionInterceptor.java
+++ b/brave-instrumentation/aws-java-sdk-v2-core/src/test/java/brave/instrumentation/awsv2/ITTracingExecutionInterceptor.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2016-2019 The OpenZipkin Authors
+ * Copyright 2016-2020 The OpenZipkin Authors
  *
  * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
  * in compliance with the License. You may obtain a copy of the License at
@@ -13,24 +13,16 @@
  */
 package brave.instrumentation.awsv2;
 
-import brave.ScopedSpan;
 import brave.SpanCustomizer;
-import brave.Tracer;
 import brave.http.HttpAdapter;
 import brave.http.HttpClientParser;
-import brave.internal.HexCodec;
-import brave.propagation.CurrentTraceContext;
-import brave.propagation.TraceContext;
 import brave.test.http.ITHttpClient;
 import java.io.IOException;
 import java.net.URI;
 import java.time.Duration;
 import java.util.Collections;
-import java.util.List;
 import java.util.concurrent.TimeUnit;
 import okhttp3.mockwebserver.MockResponse;
-import okhttp3.mockwebserver.RecordedRequest;
-import org.junit.Ignore;
 import org.junit.Test;
 import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
 import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
@@ -45,9 +37,9 @@ import software.amazon.awssdk.regions.Region;
 import software.amazon.awssdk.services.dynamodb.DynamoDbClient;
 import software.amazon.awssdk.services.dynamodb.model.GetItemRequest;
 import software.amazon.awssdk.services.dynamodb.model.PutItemRequest;
+import zipkin2.Annotation;
 import zipkin2.Span;
 
-import static java.util.Arrays.asList;
 import static org.assertj.core.api.Assertions.assertThat;
 
 public class ITTracingExecutionInterceptor extends ITHttpClient<DynamoDbClient> {
@@ -132,90 +124,13 @@ public class ITTracingExecutionInterceptor extends ITHttpClient<DynamoDbClient> 
     server.enqueue(new MockResponse());
     get(client, uri);
 
-    List<Span> spans = asList(takeSpan(), takeSpan());
-    spans.stream().filter(s -> s.parentId() != null).forEach(s -> {
-      assertThat(s.remoteServiceName())
-          .isEqualTo("dynamodb");
-      assertThat(s.name()).isEqualTo("getitem");
-
-      assertThat(s.tags())
-          .containsEntry("http.url", url(uri))
-          .containsEntry("request_customizer.is_span", "false")
-          .containsEntry("response_customizer.is_span", "false");
-    });
-  }
-
-  /** Modified to check hierarchy from parent->application->client */
-  @Override public void makesChildOfCurrentSpan() throws Exception {
-    Tracer tracer = httpTracing.tracing().tracer();
-    server.enqueue(new MockResponse());
-
-    ScopedSpan parent = tracer.startScopedSpan("test");
-    try {
-      get(client, "/foo");
-    } finally {
-      parent.finish();
-    }
-
-    List<Span> spans = asList(takeSpan(), takeSpan(), takeSpan());
-    Span applicationSpan = spans.stream()
-        .filter(s -> HexCodec.toLowerHex(parent.context().spanId()).equals(s.parentId()))
-        .findFirst()
-        .get();
-    Span clientSpan =
-        spans.stream().filter(s -> applicationSpan.id().equals(s.parentId())).findFirst().get();
-
-    RecordedRequest request = server.takeRequest();
-    assertThat(request.getHeader("x-b3-traceId"))
-        .isEqualTo(parent.context().traceIdString());
-    assertThat(request.getHeader("x-b3-parentspanid"))
-        .isEqualTo(applicationSpan.id());
-    assertThat(request.getHeader("x-b3-spanid"))
-        .isEqualTo(clientSpan.id());
-
-    assertThat(spans)
-        .extracting(Span::kind)
-        .containsOnly(null, Span.Kind.CLIENT, null);
-  }
-
-  @Override public void propagatesSpan() throws Exception {
-    super.propagatesSpan();
-    takeSpan();
-  }
-
-  @Override @Test public void propagates_sampledFalse() throws Exception {
-    server.enqueue(new MockResponse());
-    try (CurrentTraceContext.Scope ignored = httpTracing.tracing().currentTraceContext().newScope(
-        TraceContext.newBuilder().traceId(1).spanId(1).sampled(false).build())) {
-      get(client, "baz");
-    }
-
-    RecordedRequest request = server.takeRequest();
-    assertThat(request.getHeaders().toMultimap())
-        .containsKeys("x-b3-traceId", "x-b3-spanId")
-        // .doesNotContainKey("x-b3-parentSpanId") AWS SDK tracing uses 2 spans, so there is a parent
-        .containsEntry("x-b3-sampled", asList("0"));
-  }
-
-  @Override @Test public void addsErrorTagOnTransportException() throws Exception {
-    super.addsErrorTagOnTransportException();
-    Span applicationSpan = takeSpan();
-    assertThat(applicationSpan.kind()).isNull();
-    assertThat(applicationSpan.tags()).containsKey("error");
-  }
-
-  // This fails with NPE as we clear the client span on fail. on retry it isn't there anymore.
-  // we should test that the client span IDs are different and match headers sent
-  @Test @Ignore("TODO: we haven't properly addressed multiple client spans yet")
-  public void retriesAreChildrenOfApplicationSpan() throws Exception {
-    server.enqueue(new MockResponse().setResponseCode(500));
-    server.enqueue(new MockResponse());
-
-    get(client, "doo");
-
-    Span client1 = takeSpan(), client2 = takeSpan(), applicationSpan = takeSpan();
-    assertThat(applicationSpan.kind()).isNull();
-    assertThat(applicationSpan.tags()).containsKey("error");
+    Span span = takeSpan();
+    assertThat(span.remoteServiceName()).isEqualTo("dynamodb");
+    assertThat(span.name()).isEqualTo("getitem");
+    assertThat(span.tags())
+        .containsEntry("http.url", url(uri))
+        .containsEntry("request_customizer.is_span", "false")
+        .containsEntry("response_customizer.is_span", "false");
   }
 
   /** Body's inherently have a structure, and we use the operation name as the span name */
@@ -234,29 +149,22 @@ public class ITTracingExecutionInterceptor extends ITHttpClient<DynamoDbClient> 
         .isEqualTo("dynamodb");
     assertThat(span.name())
         .isEqualTo("putitem");
-
-    // application span
-    span = takeSpan();
-    assertThat(span.name()).isEqualTo("aws-sdk");
     assertThat(span.tags().get("aws.service_name"))
         .isEqualTo("DynamoDb");
     assertThat(span.tags().get("aws.operation"))
         .isEqualTo("PutItem");
   }
 
-  @Override public void propagatesExtra_newTrace() throws Exception {
-    super.propagatesExtra_newTrace();
-    takeSpan();
-  }
-
-  @Override public void reportsClientKindToZipkin() throws Exception {
-    super.reportsClientKindToZipkin();
-    takeSpan();
-  }
-
-  @Override public void reportsSpanOnTransportException() throws Exception {
-    super.reportsSpanOnTransportException();
-    takeSpan();
+  /** Make sure retry attempts are all annotated in the span. */
+  @Test public void retriesAnnotated() throws Exception {
+    server.enqueue(new MockResponse().setResponseCode(500));
+    server.enqueue(new MockResponse());
+    get(client, "/");
+    Span span = takeSpan();
+    assertThat(span.remoteServiceName()).isEqualTo("dynamodb");
+    assertThat(span.name()).isEqualTo("getitem");
+    assertThat(span.annotations()).extracting(Annotation::value)
+        .containsExactly("ws", "wr", "ws", "wr");
   }
 
   /*

--- a/brave-instrumentation/aws-java-sdk-v2-core/src/test/java/brave/instrumentation/awsv2/ITTracingExecutionInterceptor.java
+++ b/brave-instrumentation/aws-java-sdk-v2-core/src/test/java/brave/instrumentation/awsv2/ITTracingExecutionInterceptor.java
@@ -165,6 +165,12 @@ public class ITTracingExecutionInterceptor extends ITHttpClient<DynamoDbClient> 
     assertThat(span.name()).isEqualTo("getitem");
     assertThat(span.annotations()).extracting(Annotation::value)
         .containsExactly("ws", "wr", "ws", "wr");
+
+    // Ensure all requests have the injected headers.
+    assertThat(server.takeRequest().getHeader("x-b3-spanid"))
+        .isEqualTo(span.id());
+    assertThat(server.takeRequest().getHeader("x-b3-spanid"))
+        .isEqualTo(span.id());
   }
 
   /*

